File: dumper/legacy/extract/extract_triggers_frida.js
// Frida script: hook set_TriggerType to capture ability→trigger mappings
// Run after game initializes

"use strict";

const GAME_MODULE = "GameAssembly.so";

function readCString(ptr) {
    try { return ptr.isNull() ? null : ptr.readUtf8String(); } 
    catch(e) { return null; }
}

const results = {};

// Wait for game to initialize, then hook the TriggerType setter
// by finding the method via IL2CPP API
function setup() {
    const mod = Process.getModuleByName(GAME_MODULE);
    if (!mod) { console.log("[-] GameAssembly.so not found"); return; }
    
    const exports = {};
    const needed = [
        "il2cpp_class_from_name", "il2cpp_class_get_method_from_name",
        "il2cpp_method_get_pointer", "il2cpp_class_get_name",
        "il2cpp_object_get_class", "il2cpp_domain_get",
        "il2cpp_assembly_get_image", "il2cpp_domain_get_assemblies",
        "il2cpp_image_get_class_count", "il2cpp_image_get_class"
    ];
    
    for (const name of needed) {
        const addr = mod.findExportByName(name);
        if (addr) exports[name] = addr;
    }
    
    console.log("[*] Found exports:", Object.keys(exports).join(", "));
    
    // Hook il2cpp_runtime_class_init to detect when AbilityConstants type is initialized
    const classInit = mod.findExportByName("il2cpp_runtime_class_init");
    if (!classInit) { console.log("[-] class_init not found"); return; }
    
    const class_get_name = new NativeFunction(exports["il2cpp_class_get_name"], 'pointer', ['pointer']);
    
    let hooked = false;
    Interceptor.attach(classInit, {
        onEnter(args) {
            const klass = args[0];
            const name = readCString(class_get_name(klass));
            if (name === "AbilityConstants" && !hooked) {
                hooked = true;
                console.log("[+] AbilityConstants initializing! klass=", klass);
                
                // Now look for the TriggerType setter
                const setterName = Memory.allocUtf8String("set_TriggerType");
                const class_get_method = new NativeFunction(
                    exports["il2cpp_class_get_method_from_name"], 
                    'pointer', ['pointer', 'pointer', 'int']
                );
                const method_get_ptr = new NativeFunction(
                    exports["il2cpp_method_get_pointer"],
                    'pointer', ['pointer']
                );
                const obj_get_class = new NativeFunction(
                    exports["il2cpp_object_get_class"],
                    'pointer', ['pointer']
                );
                
                // Find Ability base class
                const nsName = Memory.allocUtf8String("Spacewood.Core.Models.Abilities");
                const abilityName = Memory.allocUtf8String("Ability");
                const class_from_name = new NativeFunction(
                    exports["il2cpp_class_from_name"],
                    'pointer', ['pointer', 'pointer', 'pointer']
                );
                
                // Search all images
                const domain = new NativeFunction(exports["il2cpp_domain_get"], 'pointer', [])();
                const getAssemblies = new NativeFunction(exports["il2cpp_domain_get_assemblies"], 'pointer', ['pointer', 'pointer']);
                const sizeRef = Memory.alloc(4);
                const asmArray = getAssemblies(domain, sizeRef);
                const count = sizeRef.readU32();
                console.log(`[*] ${count} assemblies`);
                
                const asmGet = new NativeFunction(exports["il2cpp_assembly_get_image"], 'pointer', ['pointer']);
                const imgGetCount = new NativeFunction(exports["il2cpp_image_get_class_count"], 'uint', ['pointer']);
                const imgGetClass = new NativeFunction(exports["il2cpp_image_get_class"], 'pointer', ['pointer', 'uint']);
                
                let abilityKlass = null;
                
                for (let i = 0; i < count; i++) {
                    const asm = asmArray.add(i * 8).readPointer();
                    const img = asmGet(asm);
                    const classCount = imgGetCount(img);
                    
                    for (let j = 0; j < classCount; j++) {
                        const k = imgGetClass(img, j);
                        const n = readCString(class_get_name(k));
                        if (n === "Ability") {
                            abilityKlass = k;
                            console.log(`[+] Found Ability class at ${k}`);
                            break;
                        }
                    }
                    if (abilityKlass) break;
                }
                
                if (!abilityKlass) {
                    console.log("[-] Ability class not found");
                    return;
                }
                
                // Get set_TriggerType method
                const method = class_get_method(abilityKlass, setterName, 1);
                if (!method.isNull()) {
                    const funcPtr = method_get_ptr(method);
                    console.log(`[+] set_TriggerType at ${funcPtr}`);
                    
                    // Hook it!
                    Interceptor.attach(funcPtr, {
                        onEnter(args) {
                            const thisObj = args[0];  // 'this' pointer
                            const triggerVal = args[1].toInt32();  // TriggerEvent enum value
                            const klass = obj_get_class(thisObj);
                            const className = readCString(class_get_name(klass));
                            if (className && className.endsWith("Ability")) {
                                results[className] = triggerVal;
                                console.log(`[*] ${className} -> TriggerEvent[${triggerVal}]`);
                            }
                        }
                    });
                }
            }
        }
    });
}

// Periodically dump results
setInterval(() => {
    if (Object.keys(results).length > 0) {
        console.log("=== RESULTS ===");
        console.log(JSON.stringify(results, null, 2));
    }
}, 5000);

// Start setup when script loads
setTimeout(setup, 100);

File: dumper/legacy/extract/extract_triggers_frida.py
"""
SAP trigger extractor — deferred read approach.

SetTrigger(TriggerMinions) is called while the TriggerMinions object still has
TriggerType=0. The builder sets TriggerType AFTER storing the ref. So we save
every (ability_ptr, triggerMinions_ptr) pair during SetTrigger calls, then
on Ctrl-C read the CURRENT values from those (now-fully-initialised) objects.

Ability names come from UnityEngine.Object.get_name() via il2cpp_runtime_invoke.
Trigger type comes from il2cpp_object_get_class → il2cpp_class_get_name on the
TriggerMinions object (class name = trigger type, e.g. TriggerSell, TriggerDeath).
"""
import frida, json, signal, sys, os, time
from pathlib import Path

GAME_DIR = str(Path.home() / ".local/share/Steam/steamapps/common/Super Auto Pets")
GAME_BIN = f"{GAME_DIR}/superautopets.x86_64"
OUTPUT   = str(Path.home() / "Documents/GitHub/sap-data-scrape/trigger_mapping_frida.json")

JS = r"""
"use strict";

function cstr(p) {
    try { return (p && !p.isNull()) ? p.readUtf8String() : null; }
    catch(e) { return null; }
}
function readCsString(p) {
    try {
        if (!p || p.isNull()) return null;
        const len = p.add(12).readU32();
        if (len === 0 || len > 4096) return null;
        return p.add(16).readUtf16String(len);
    } catch(e) { return null; }
}

// State populated during hooks, read later on dump.
const savedAbilities = new Map();  // self_ptr_str → {self}
const hookNames      = new Map();  // self_ptr_str → locoKey (from SetAbout hook)

// Globals bound in setupAfterInit, used in dumpAll
let _objectGetClass  = null;
let _classGetName    = null;
let _classGetFields  = null;
let _fieldGetName    = null;
let _fieldStaticGetV = null;
let _abilityEnumClass = null;

// ── dump: called when Python sends {type:"dump"} ──────────────────────────
function dumpAll() {
    // Build AbilityEnum int→name map (deferred — statics initialized by now)
    const abilityEnumNames = {};
    if (_abilityEnumClass && _classGetFields && _fieldGetName && _fieldStaticGetV) {
        const it = Memory.alloc(Process.pointerSize); it.writePointer(ptr(0));
        let fld;
        while (!(fld = _classGetFields(_abilityEnumClass, it)).isNull()) {
            const fn_ = cstr(_fieldGetName(fld));
            if (fn_ && fn_ !== "value__") {
                try {
                    const buf = Memory.alloc(4);
                    _fieldStaticGetV(fld, buf);
                    abilityEnumNames[buf.readS32()] = fn_;
                } catch(e) {}
            }
        }
        send({t:"log", m:`AbilityEnum: ${Object.keys(abilityEnumNames).length} values`});
        const sample = Object.entries(abilityEnumNames).slice(0,5).map(([k,v])=>`${k}=${v}`).join(", ");
        send({t:"log", m:`  sample: ${sample}`});
    }

    const results = [];
    for (const [key, {self}] of savedAbilities) {
        // 1. Name from SetAbout hook (most reliable)
        let abilityName = hookNames.get(key) || null;

        // 2. Name from AbilityEnum field @ offset 16
        if (!abilityName && Object.keys(abilityEnumNames).length > 0) {
            try {
                const enumInt = self.add(16).readS32();
                abilityName = abilityEnumNames[enumInt] || null;
            } catch(e) {}
        }

        // 3. Raw enum int for diagnostics
        let enumInt = -1;
        try { enumInt = self.add(16).readS32(); } catch(e) {}

        // 4. Fallback: AboutLocoKey @ offset 40
        if (!abilityName) {
            try {
                const strPtr = self.add(40).readPointer();
                abilityName = readCsString(strPtr);
            } catch(e) {}
        }

        // Get trigger class name from TriggerMinions object at offset 440
        let trigPtr = null, trigClassName = null;
        try { trigPtr = self.add(440).readPointer(); } catch(e) {}
        if (trigPtr && !trigPtr.isNull() && _objectGetClass && _classGetName) {
            try {
                const tc = _objectGetClass(trigPtr);
                trigClassName = cstr(_classGetName(tc));
            } catch(e) {}
        }

        results.push({
            name:     abilityName,
            enum_int: enumInt,
            trig_cls: trigClassName,
            self:     key,
        });
    }
    send({t:"dump_result", data: results, count: results.length});
}

recv("dump", (_msg) => { dumpAll(); });

function setupAfterInit(mod) {
    function nfn(name, ret, args) {
        const a = mod.findExportByName(name);
        if (!a) { send({t:"log", m:"Missing export: " + name}); return null; }
        return new NativeFunction(a, ret, args);
    }

    const domainGet           = nfn("il2cpp_domain_get",            "pointer", []);
    const domainGetAssemblies = nfn("il2cpp_domain_get_assemblies", "pointer", ["pointer","pointer"]);
    const imageGetClassCount  = nfn("il2cpp_image_get_class_count", "uint32",  ["pointer"]);
    const imageGetClass       = nfn("il2cpp_image_get_class",       "pointer", ["pointer","uint32"]);
    const classGetName        = nfn("il2cpp_class_get_name",        "pointer", ["pointer"]);
    const classGetNamespace   = nfn("il2cpp_class_get_namespace",   "pointer", ["pointer"]);
    const classGetMethods     = nfn("il2cpp_class_get_methods",     "pointer", ["pointer","pointer"]);
    const classGetFields      = nfn("il2cpp_class_get_fields",      "pointer", ["pointer","pointer"]);
    const fieldGetName        = nfn("il2cpp_field_get_name",        "pointer", ["pointer"]);
    const fieldStaticGetValue = nfn("il2cpp_field_static_get_value","void",    ["pointer","pointer"]);
    const methodGetName       = nfn("il2cpp_method_get_name",       "pointer", ["pointer"]);
    const objectGetClass      = nfn("il2cpp_object_get_class",      "pointer", ["pointer"]);

    _objectGetClass  = objectGetClass;
    _classGetName    = classGetName;
    _classGetFields  = classGetFields;
    _fieldGetName    = fieldGetName;
    _fieldStaticGetV = fieldStaticGetValue;

    if (!domainGet) return;

    const domain   = domainGet();
    const sizePtr  = Memory.alloc(8); sizePtr.writeU64(0);
    const asmArray = domainGetAssemblies(domain, sizePtr);
    const asmCount = sizePtr.readU32();
    send({t:"log", m:`${asmCount} assemblies found`});

    const assemblyGetImage = nfn("il2cpp_assembly_get_image", "pointer", ["pointer"]);

    let abilityClass = null;

    for (let ai = 0; ai < asmCount; ai++) {
        const asmPtr = asmArray.add(ai * Process.pointerSize).readPointer();
        const img = assemblyGetImage(asmPtr);
        const n = imageGetClassCount(img);
        for (let ci = 0; ci < n; ci++) {
            const klass = imageGetClass(img, ci);
            if (!klass || klass.isNull()) continue;
            const cname = cstr(classGetName(klass));
            if (!cname) continue;

            if (cname === "Ability" && !abilityClass) {
                abilityClass = klass;
                send({t:"log", m:"Found Ability class"});
            }
            if (cname === "AbilityEnum" && !_abilityEnumClass) {
                _abilityEnumClass = klass;
                send({t:"log", m:"Found AbilityEnum class"});
            }
            if (abilityClass && _abilityEnumClass) break;
        }
        if (abilityClass && _abilityEnumClass) break;
    }

    if (!abilityClass)     { send({t:"log", m:"ERROR: Ability class not found"}); return; }
    if (!_abilityEnumClass){ send({t:"log", m:"WARNING: AbilityEnum not found — will use SetAbout hook only"}); }

    // Find and hook SetTrigger, SetAimAndTrigger, SetAbout
    const toHook = {SetTrigger: null, SetAimAndTrigger: null, SetAbout: null};
    {
        const it = Memory.alloc(Process.pointerSize); it.writePointer(ptr(0));
        let mth;
        while (!(mth = classGetMethods(abilityClass, it)).isNull()) {
            const mn = cstr(methodGetName(mth));
            if (mn && toHook.hasOwnProperty(mn) && !toHook[mn]) {
                toHook[mn] = mth.readPointer();
            }
        }
    }

    let hookedCount = 0;

    // SetAbout(string locoKey) — captures ability name→ptr mapping
    const setAboutPtr = toHook["SetAbout"];
    if (setAboutPtr && !setAboutPtr.isNull()) {
        try {
            Interceptor.attach(setAboutPtr, {
                onEnter(args) {
                    const s = readCsString(args[1]);
                    if (s) {
                        // locoKey format: "Ability.BeaverAbility.1.About" or just "BeaverAbility"
                        // Extract ability name from key
                        let name = s;
                        const parts = s.split('.');
                        if (parts.length >= 2 && parts[0] === 'Ability') name = parts[1];
                        hookNames.set(args[0].toString(), name);
                    }
                }
            });
            hookedCount++;
            send({t:"log", m:"Hooked SetAbout (for ability names)"});
        } catch(e) {
            send({t:"log", m:`SetAbout hook failed: ${e.message}`});
        }
    } else {
        send({t:"log", m:"SetAbout: not found"});
    }

    for (const label of ["SetTrigger", "SetAimAndTrigger"]) {
        const fnPtr = toHook[label];
        if (!fnPtr || fnPtr.isNull()) { send({t:"log", m:`${label}: not found`}); continue; }
        try {
            Interceptor.attach(fnPtr, {
                onEnter(args) {
                    savedAbilities.set(args[0].toString(), {self: args[0]});
                }
            });
            hookedCount++;
            send({t:"log", m:`Hooked ${label}`});
        } catch(e) {
            send({t:"log", m:`${label} hook failed: ${e.message}`});
        }
    }

    send({t:"log", m:`Ready. ${hookedCount} hooks active. Ctrl-C after game loads to dump.`});
}

const pollTimer = setInterval(() => {
    const mod = Process.findModuleByName("GameAssembly.so");
    if (!mod) return;
    clearInterval(pollTimer);
    send({t:"log", m:`GameAssembly.so @ ${mod.base}`});
    try { setupAfterInit(mod); }
    catch(e) { send({t:"log", m:"Error: " + e.message + "\n" + e.stack}); }
}, 100);
"""

saved_count = 0
dump_result = None


def on_message(msg, _data):
    global saved_count, dump_result
    if msg['type'] == 'send':
        p = msg['payload']
        t = p.get('t', '')
        if t == 'log':
            print(f"  [JS] {p['m']}", flush=True)
        elif t == 'dump_result':
            dump_result = p['data']
            print(f"  [*] Dump received: {p['count']} entries", flush=True)
        elif t == 'err':
            print(f"  [JS-ERR] {p['msg']}", flush=True)
    elif msg['type'] == 'error':
        print(f"  [FRIDA-ERR] {msg.get('description','?')}", flush=True)


def build_output(dump_result):
    """
    Collapse 3-level duplicates: each ability appears 3× (one per level) with
    identical trigger. Keep one entry per unique ability name.
    Output: { "BeaverAbility": "TriggerSell", ... }
    """
    seen = {}
    for entry in dump_result:
        name = entry.get('name')
        trig = entry.get('trig_cls')
        if not name:
            name = f"unknown_{entry.get('self','')[-6:]}"
        if name not in seen:
            seen[name] = trig or '?'
    return seen


def save_and_exit(pid, scr):
    print("\n[*] Requesting dump from JS...", flush=True)
    scr.post({"type": "dump"})
    for _ in range(30):
        time.sleep(0.2)
        if dump_result is not None:
            break
    else:
        print("[!] Dump timed out")

    mapping = build_output(dump_result or [])
    print(f"\n[*] {len(mapping)} unique ability→trigger mappings:")
    for k, v in sorted(mapping.items()):
        print(f"  {k:45s} -> {v}")

    with open(OUTPUT, 'w') as f:
        json.dump({"trigger_map": mapping, "raw_dump": dump_result}, f, indent=2)
    print(f"\n[*] Saved to {OUTPUT}")

    try:
        import subprocess
        subprocess.run(["kill", str(pid)], capture_output=True)
    except Exception:
        pass


def main():
    env = {**os.environ, "DISPLAY": ":0"}
    print("[*] Spawning SAP...")
    pid = frida.spawn([GAME_BIN], cwd=GAME_DIR, env=env)
    print(f"[*] PID {pid}")

    sess = frida.attach(pid)
    scr  = sess.create_script(JS)
    scr.on('message', on_message)
    scr.load()
    frida.resume(pid)

    print("[*] Hooks active. Wait for game to load fully, then press Ctrl-C.")

    def handle_sigint(_sig, _frame):
        save_and_exit(pid, scr)
        sys.exit(0)

    signal.signal(signal.SIGINT, handle_sigint)
    while True:
        time.sleep(1)


if __name__ == "__main__":
    main()
